### DISCUSSION TITLE
Node based logic editor?

### DISCUSSION THREAD

**Discussion by tigert** [MobiFlight/MobiFlight-Connector#1277] [Ideas]
As a less of a programmer, I am wondering if a node-based editor for cockpit event referencing and transforms etc would be a viable approach?

This screenshot of Blender's game engine logic editor plugin is one example how it could look like:

![image](https://github.com/MobiFlight/MobiFlight-Connector/assets/2587818/7be51caf-2135-480d-b17b-a590cfd99ca1)

This could visualize config references pretty easily, by adding other config items as nodes, and transforms could be added between the links too.

I am wondering actually could we do the whole UI like this, by having a default "input" node that has an input device as one node, and a simvar as another, and then anything in the notebook tabs for transform etc could be added as nodes that go between the input and simvar.

Same for outputs, you could have the simvar as a node that outputs data, and it gets connected to an output device node by default. And if you wanted, you could pull in other objects as references,  have logic AND/OR nodes,

I kinda think this all could work as nodes - and the most common case could be represented then as a single config line in a list, and if the item contains more complex logic, it could have a "edit the logic" button instead of just an Encoder -> Simvar arrow.

**Comment by MobiFlight-Admin** [comment]
I like the idea - i have thought about it many times. I know it from Blender and from nodeRED - there are a few open source libraries, but i will have to change the UI framework first. 

On the other hand I am not sure how well beginners really feel about it.

**Comment by tigert** [comment]
I don't know if this would lead to some dead end in the UI, but the more I think of this, the more I like it.

I agree with the beginner thing - but maybe we can kinda make the default nodes look less complex?

Something like this?
![image](https://github.com/MobiFlight/MobiFlight-Connector/assets/2587818/a005e71c-1ddc-4711-b8b4-7a265918f538)

But it would really need to have a nice, non-in-your-face transition to edit the node contents (like on the above the input device selection needs to be thought out very well etc...)

**Comment by tigert** [comment]
Something like this for an initial view, then clicking the "scan for input" or selecting the device from dropdown or similar controls:

![image](https://github.com/MobiFlight/MobiFlight-Connector/assets/2587818/f911afc8-1377-452f-8c71-2f2d1c1961ed)

We are left with a fairly simple node setup. We probably should have blue buttons for those things that you should click to set things up, guiding the user a bit , like first scan the input, then selecting the sim events.

But we could do a bit of a user testing to see if folks get the idea - there are ways to do interactive mockups that have a simple config UI flow and we can see if folks get the idea how it works.

**Comment by tigert** [comment]
I guess I am thinking visual nodes because I find some of the more advanced logical tasks difficult to do when the things are scattered in many places accross the UI.

One idea would be to have the inputs and outputs as simple mappings: Input device to sim event, and simvar to output device - I would guess most of the config items will be like this. But then you could have more complex node constructions where you use MF variables or the other config items via a config reference node, and could structure your logic in a separate large "canvas" where you can see the whole flow at once.

One alternative idea is to just run Node-RED and create plugins for data input and output from MobiFlight using some web api.
